Simulator/simulator.py

```diff
@@ -95,324 +85,687 @@ def __init__(self, etype: etype, time: datetime, job_id: int):
-      (self.input_trace.job_id >= h * self.job_count)
-      & (self.input_trace.job_id <= (1.0 - h) * self.job_count)
-      ]
-    mean_wait_time_sec = steady_state[
-      steady_state.schedule == "R"
-      ].wait_time_sec.mean()
-    balking_rate = (
-            1
-            - self.input_trace[self.input_trace.schedule == "R"].time.count()
-            / self.input_trace.time.count()
-    )
-
-    # LJW and Compound policy -- On-demand jobs don't wait in queue
-    mean_wait_time_sec = mean_wait_time_sec * (1 - balking_rate)
-
-    # Simulation time
-    elapsed_simulation_time = (end - start).total_seconds()
-
-    # User has not specified trace duration
-    if duration_hrs is None:
-      duration_hrs = elapsed_simulation_time / 3600
-
-    # Computing total cost
-    (
-      total_cost,
-      reserved_cost,
-      on_demand_cost,
-      effective_price,
-    ) = self._compute_total_cost(self.input_trace, duration_hrs)
-
-    print(
-      "Reserved cost is ",
-      reserved_cost,
-      ", On demand cost is ",
-      on_demand_cost,
-      ", Total cost is ",
-      reserved_cost + on_demand_cost,
-      )
-    print("----------------------------------------------------------------")
-
-    return (mean_wait_time_sec, effective_price, balking_rate, total_cost)
+                break
+
+
+    def resubmit_long_jobs_after_spec_execution(self, job, runtime_threshold_sec):
+        # Resubmit job after running for "t" time per speculative execution
+        job.loc['speculative_execution'] = True
+        job.loc['submit_time'] = job['submit_time'] + runtime_threshold_sec
+
+        # Place job in correct place in trace
+        temp_df = self.input_trace[self.input_trace['submit_time'] > job.submit_time]
+        new_index = temp_df.iloc[0].job_id
+        job.loc['job_id'] = new_index
+
+        job = job.to_frame().T
+
+        # Reorder and re-index input trace to reflect new timing of job
+        self.input_trace = pd.concat([self.input_trace.iloc[:new_index], job,
+                                      self.input_trace.iloc[new_index:]]).reset_index(drop=True)
+        self.input_trace['job_id'] = self.input_trace.index
+
+    def calc_on_demand_cost_for_job(self, job, runtime):
+        CPUs_needed = self.VM_CPU * math.ceil(job.allocated_CPUs / self.VM_CPU)
+        return CPUs_needed * (runtime / SECONDS_IN_HOUR) * UNIT_CPU_COST_HR
+
+    def print_results(self) -> None:
+        print(
+            "Reserved cost is ",
+            self.reserved_cost,
+            ", On demand cost is ",
+            self.on_demand_cost,
+            ", Total cost is ",
+            self.reserved_cost + self.on_demand_cost,
+            ", Mean waiting time is ",
+            self.mean_waiting_time / 60, " minutes"
+        )
+
+    """
+    Train waiting time model on training trace given specified waiting time threshold; return model fit on training data
+    """
+
+    def train_waiting_time_model(self, waiting_time_threshold_sec: int):
```

Comment:
nit: should we consider moving to a different class?

Reply:
good call - definitely cleaner. reflected!